internal/server/v2/observation/facet.go

```diff
@@ -0,0 +1,219 @@
+	for _, btData := range facetBtDataList {
+		for _, row := range btData {
+			sv := row.Parts[variableRowPart]
+			if _, ok := svToFacetList[sv]; !ok {
+				svToFacetList[sv] = []*pb.Facet{}
+			}
+			svToFacetList[sv] = append(svToFacetList[sv], row.Data.(*pb.Facets).GetFacets()...)
+		}
+	}
+	// Go through each list of facets, sort and remove duplicates, and add to
+	// result.
+	for sv, facetList := range svToFacetList {
+		entityObservation := &pbv2.EntityObservation{}
+		sort.Sort(ranking.FacetByRank(facetList))
+		seenFacets := map[string]string{}
+		for _, facet := range facetList {
+			facetID := util.GetFacetID(facet)
+			if _, ok := seenFacets[facetID]; ok {
+				continue
+			}
+			seenFacets[facetID] = ""
+			entityObservation.OrderedFacets = append(entityObservation.OrderedFacets, &pbv2.FacetObservation{FacetId: facetID})
+			result.Facets[facetID] = facet
+		}
+		result.ByVariable[sv].ByEntity[""] = entityObservation
+	}
+	return result
+}
+
+// ContainedInFacet implements logic to get the available facets for child
+// places contained in an ancestor place for a list of variables.
+func ContainedInFacet(
+	ctx context.Context,
+	store *store.Store,
+	cache *resource.Cache,
+	metadata *resource.Metadata,
+	httpClient *http.Client,
+	remoteMixer string,
+	variables []string,
+	ancestor string,
+	childType string,
+	queryDate string,
+) (*pbv2.ObservationResponse, error) {
+	result := &pbv2.ObservationResponse{
+		ByVariable: map[string]*pbv2.VariableObservation{},
+		Facets:     map[string]*pb.Facet{},
+	}
+	for _, variable := range variables {
+		result.ByVariable[variable] = &pbv2.VariableObservation{
+			ByEntity: map[string]*pbv2.EntityObservation{},
+		}
+		result.ByVariable[variable].ByEntity[""] = &pbv2.EntityObservation{}
+	}
+	if store.BtGroup != nil {
+		readCollectionCache := hasCollectionCache(ancestor, childType)
+		if readCollectionCache && queryDate != "" {
+			btDataList, err := bigtable.Read(
+				ctx,
+				store.BtGroup,
+				bigtable.BtObsCollectionFacet,
+				[][]string{{ancestor}, {childType}, variables, {queryDate}},
+				func(jsonRaw []byte) (interface{}, error) {
+					var facets pb.Facets
+					if err := proto.Unmarshal(jsonRaw, &facets); err != nil {
+						return nil, err
+					}
+					return &facets, nil
+				},
+			)
+			if err != nil {
+				return nil, err
+			}
+			result = GetFacetObsResponse(variables, btDataList, 2)
+		} else {
+			childPlaces, err := fetchChildPlaces(
+				ctx, store, metadata, httpClient, remoteMixer, ancestor, childType)
+			if err != nil {
+				return nil, err
+			}
+			totalSeries := len(variables) * len(childPlaces)
+			if totalSeries > maxSeries {
+				return nil, status.Errorf(
+					codes.Internal,
+					"Stop processing large number of concurrent observation series: %d",
+					totalSeries,
+				)
+			}
+			log.Println("Fetch series cache in contained-in observation query")
+			btData, err := stat.ReadStatsPb(ctx, store.BtGroup, childPlaces, variables)
```

Comment:
should here call SeriesFacet()?

Reply:
SeriesFacet won't work if there is a specific query date. Updated to call SeriesFacet if no specified date.